### DISCUSSION TITLE
Build Error on 20241231 Version

### DISCUSSION THREAD

**Discussion by Ken-Kauffmann** [arnaud-j4k08/PicoClockGreenEasy#8] [Q&A]
Wow, you have been busy.  I tried to build your newest version, and got an error with Gps.h

'time_t' was not declared in this scope

I got it to build by adding  #include <time.h>, but I am not at all sure as that is what you intended.  I would not be at all surprised if my Visual Studio Code environment is incorrect.  I am not sure where pico/time.h is supposed to come from.  I tried creating a new project in Visual Studio Code, using the Raspberry Pi Pico Project extension, and copying over the files included in the downloaded zip file.  I originally got the build error there.  Thinking that I might have left something out, I used the "New Project From Example" from the Raspberry Pi Pico Project extension, and did much the same thing.  I copied over the src files, and modified the other files to look like yours.  

It also looks as though USA is not a valid argument in UserConfig.cmake, so I am currently using Europe
My time initializes to 12:00 on 1/1/1970

I have an Alarm Menu
I have Countdown
I have StopWatch
I have Clock Sync
I have Options

Within Options, I have AutoScroll, Time Format, Date Format, Hourly Chime, Auto Light, Brightness

Within Clock Sync, I have Source, Sync Now, Wifi, Exit
SyncNow sets the F and C characters, but does not appear to sync  (at least with RTC).  

I can update Date and Time manually, and they are stored in the RTC.  Autoscroll acts as though it does not see the RTC, no temperature.

I connected a GPS, and waited until the red led indicated that it synced to the satellites.   I manually set the date and time incorrectly.  I then did the GPS sync. and it worked correctly.  Yay!

I set the date and time incorrectly, and then synced to NTP.  It worked correctly. 

Sync to RTC does not work as expected.  the F and C icons come on, but never stop.  No matter what button is pressed. 

Your updates are looking good.  I am interested in the Gps.h error, though.  Thanks, Ken

**Comment by Ken-Kauffmann** [comment]
Some questions regarding behavior.
1)  When I unplug, and plug the clock back in,  The display goes to a time of 12:00, the day of the week indicator shows the wrong day, and the F and C indicators are on.   Within about 3 seconds, the correct time is shown, the correct day of the week is shown, the F and C indicators go away.   I am assuming that we are loading from the RTC at this stage, correct?

2) UserConfig.cmake accepts Usa, not USA.  You do not appear to be checking for either in DaylightSavingTime, but default to it if it is not Unknown or Europe.  Correct?

3) Temperature no longer appears, either scrolling through the menu entries, or in the root menu AutoScroll.  If I understand correctly, this means that m_clock.rtc() is a nullptr, and/or isAvailable() is not set.  I scanned the source changes and did not see any obvious thing that changed. 

4)  I added the WifiStatus function back into my copy of the root menu.  It appears that the status is Down, until I try to sync to NTP.  It appears that you no longer automatically try to use NTP during start up.   Is that correct?

5)  What is the Sync from RTC supposed to do?  If I manually set the time to an hour earlier, and then do the Sync Now after setting the sync source to RTC, I see the following.. The F and C icons blink.  They do not stop blinking.  If I unplug the clock, and plug it back in, I get a time of 12:00, the F and C blinking, and the date resets to January 1, 1970.   Luckily, re-syncing with GPS works like a charm, and I can get everything back.  

Just for your information, I was able to recreate my Hamclock submenu, and submenu autoscroll in the new version.  It is very easy the second time.  :-)

**Reply by arnaud-j4k08** [reply]
Hello Ken. First, let me wish you a Happy New Year.

Most of the issues you encountered have the same cause: for testing, I temporarily disabled the RTC using a preprocessor symbol called USE_RTC that I configured in UserConfig.cmake. But I did not realize that the RTC will be disabled by default if this symbol is not defined at all. I have now pushed a fix. Sorry for that.

> Wow, you have been busy. I tried to build your newest version, and got an error with Gps.h
>
>'time_t' was not declared in this scope
>
>I got it to build by adding #include <time.h>, but I am not at all sure as that is what you intended. I would not be at all surprised if my Visual Studio Code environment is incorrect. I am not sure where pico/time.h is supposed to come from. I tried creating a new project in Visual Studio Code, using the Raspberry Pi Pico Project extension, and copying over the files included in the downloaded zip file. I originally got the build error there. Thinking that I might have left something out, I used the "New Project From Example" from the Raspberry Pi Pico Project extension, and did much the same thing. I copied over the src files, and modified the other files to look like yours.

You may be using a different version of the Pico SDK that have slightly different header dependencies. I use version 1.5.1. All system headers like time.h and pico/time.h (those are different files) come from it. I have pushed a fix for that too.

>Sync to RTC does not work as expected. the F and C icons come on, but never stop. No matter what button is pressed.

Yes, this was an additional issue when running with disabled RTC. The code still started a synchronization from the RTC, although it does not make sense. I fixed it.

>When I unplug, and plug the clock back in, The display goes to a time of 12:00, the day of the week indicator shows the wrong day, and the F and C indicators are on. Within about 3 seconds, the correct time is shown, the correct day of the week is shown, the F and C indicators go away. I am assuming that we are loading from the RTC at this stage, correct?

No, RTC synchronization takes less than one second, so in this case, you were using either NTP or GPS, depending on what you selected in the "Source" menu entry.

>UserConfig.cmake accepts Usa, not USA. You do not appear to be checking for either in DaylightSavingTime, but default to it if it is not Unknown or Europe. Correct?

I called it "Usa" to match my usual naming convention in enums, but have now added "USA" as alias to avoid this confusion. Indeed, if it is not Unknown or Europe, the `DaylightSavingTime::determineDstStartAndEnd` method will default to USA. I think this method deserves to be refactored, but I decided to leave it as it is as long as only 2 DST types are supported.

>Temperature no longer appears, either scrolling through the menu entries, or in the root menu AutoScroll. If I understand correctly, this means that m_clock.rtc() is a nullptr, and/or isAvailable() is not set. I scanned the source changes and did not see any obvious thing that changed.

This is a consequence of the disabled RTC, which was done by not initializing the m_rtc unique_ptr in `Clock::Clock`. 

>I added the WifiStatus function back into my copy of the root menu. It appears that the status is Down, until I try to sync to NTP. It appears that you no longer automatically try to use NTP during start up. Is that correct?

I am not sure this is correct. If "NTP" is set as source, at the next start up, the clock will run the NTP synchronization. After changing the source, you need to wait one minute for this setting to be persisted in the flash memory. What's correct however, is that now, the wi-fi connection will only be open if a NTP synchronization is to be performed. If the sync source is set to RTC or GPS, it will not open any wi-fi connection, as none is needed.

>What is the Sync from RTC supposed to do? If I manually set the time to an hour earlier, and then do the Sync Now after setting the sync source to RTC, I see the following.. The F and C icons blink. They do not stop blinking. If I unplug the clock, and plug it back in, I get a time of 12:00, the F and C blinking, and the date resets to January 1, 1970. Luckily, re-syncing with GPS works like a charm, and I can get everything back.

The sync from RTC recovers the date/time at start-up using the backup battery. This was another consequence of the disabled RTC and is fixed now.

**Reply by Ken-Kauffmann** [reply]
Happy New Year to you as well.  

That was amazing.  Yes, removing the "#If USE_RTC " in Clock.cpp made all the diference.  I have the Temperature back, and the RTC Sync works quickly and correctly.   Thanks for the reminder about waiting 1 minute before trying the unplug/replug test to check the synchronization.   Both RTC and GPS startup synchronization are almost immediate.  

The NTP synchronization takes longer, and does not always work.  I think that this is a function of my Wifi environment, and not your code.   I need to use two different SSIDs, depending upon where I am in the house, and coverage is pretty sketchy on the one that I am currently using. 

The GPS synchronization is fantastic.  It is fast, and it is reliable.  Very impressive.  As an aside, I am using the following active antenna that I purchased from Amazon last year:  Bingfu Waterproof Active GPS Navigation Antenna Adhesive Mount SMA Male GPS Antenna with 15cm 6 inch U.FL IPX IPEX to SMA Female RG178 Coaxial Pigtail Cable for GPS Module Receiver Tracking.  It comes with an adapter that clips onto the plug on the GPS.  GPS sync ( blinking red LED) is almost immediate with it.   I find that interference with Bluetooth, and USB 3.0 has the biggest effect on getting a lock, so I like to put the antenna away from the PC.  

In my version of Visual Studio Code, the Pico SDK version reads 2.1.0.  Maybe that does make a difference with the header dependencies. 

One last question ( for today at least )  :-)  

In my UserConfig.cmake file, I have "set(TRACE_TO_STDIO "1")".   Trace.cpp contains: 
bool isEnabledForFile(const std::string &file)
{
    return 
        file.find("Clock.cpp") != std::string::npos ||
        file.find("main.cpp") != std::string::npos ||
        file.find("Rtc.cpp") != std::string::npos;
}

The size of my .uf2 indicates that tracing has been included in the build.    Device Manager in Windows shows that the Clock is connected as COM4.  I am not getting any trace in the terminal window in Visual Studio Code.  Have I forgotten something?

**Reply by Ken-Kauffmann** [reply]
Never mind with the Trace questions.  I used Tera Term, and have all the traces that I could want.   Dopey me.

**Reply by arnaud-j4k08** [reply]
Great that it works for you too now.

>The NTP synchronization takes longer, and does not always work. I think that this is a function of my Wifi environment, and not your code. I need to use two different SSIDs, depending upon where I am in the house, and coverage is pretty sketchy on the one that I am currently using.

With the last changes, there are no timeout for the wifi anymore. It tries connecting until it succeeds. Actually, the wifi router is also barely reachable in certain rooms of my house, so I will check what happens there.

>The GPS synchronization is fantastic. It is fast, and it is reliable. Very impressive. As an aside, I am using the following active antenna that I purchased from Amazon last year: Bingfu Waterproof Active GPS Navigation Antenna Adhesive Mount SMA Male GPS Antenna with 15cm 6 inch U.FL IPX IPEX to SMA Female RG178 Coaxial Pigtail Cable for GPS Module Receiver Tracking. It comes with an adapter that clips onto the plug on the GPS. GPS sync ( blinking red LED) is almost immediate with it. I find that interference with Bluetooth, and USB 3.0 has the biggest effect on getting a lock, so I like to put the antenna away from the PC.

I am glad you like it. Originally, I was looking for a way to synchronize my Pico (the one without W) and my first idea was to use DCF77 (equivalent to the american WWVB). I wasted a lot of time trying to get it to work, to finally find that it is too sensitive to interferences. So I decided to use GPS instead, which is way more reliable and actually easier to implement.

Thanks for the tip about the active antenna. I had briefly looked at some, but I was not sure if they can work with such a GPS module.

>In my version of Visual Studio Code, the Pico SDK version reads 2.1.0. Maybe that does make a difference with the header dependencies.

Yes, that must be the explanation. That's also a new task for me :-) : update my Pico SDK to the latest version

>Never mind with the Trace questions. I used Tera Term, and have all the traces that I could want. Dopey me.

I had a similar experience. At the beginning, I was also using the serial monitor in Visual Studio Code, but it was loosing the connection each time I restarted the Pico. Then I found Tera Term, which is a lot more convenient.

By the way, did you notice that you can control PicoClockGreenEasy from the terminal? The Enter, Arrow up and Arrow down keys simulate the 3 buttons of the Pico Clock Green.

**Reply by Ken-Kauffmann** [reply]
I had seen the code block in ClockUi.cpp, and the SIMULATE_BUTTONS_FROM_STDIO in the UserConfig.cmake file, but had never tried it until you suggested it up above.   And, it even works if you aren't tracing anything.   Very nice.
